### PR TITLE
New fields in material: z retract = z distance before contact, Z feed before contact, hard crust thikness, hard crust Z feed in%. Endmill: new element endmill shape Disk, new field end mill thikness (useful for threaded with shape disk). Scale in x, y and/or z direction

### PR DESCRIPTION
**Pull request by MARIOBASZ**
Observation: I just want to upload the scale icon and the scaling plugin. I do not know how to do so that the other changes listed do not appear

### CODE REVIEW

**Comment by MARIOBASZ** [comment]
scale in xy plane 2:1
image

scale 2:1 in x direction, 1.5:1 in y direction
image

First version, lack correct

**Comment by sfinexer** [comment]
@MARIOBASZ 
Where is the PNG file scale?

sfinexer@sfinexer:/media/files/выжигание/bCNC-master$ ./bCNC.py
Traceback (most recent call last):
  File "/media/files/выжигание/bCNC-master/ToolsPage.py", line 934, in __init__
    exec("import %s"%(name))
  File "<string>", line 1, in <module>
  File "/media/files/выжигание/bCNC-master/plugins/slicemesh.py", line 32, in <module>
    import stl #FIXME: write smaller STL parser
  File "/media/files/выжигание/bCNC-master/lib/stl/__init__.py", line 1, in <module>
    from .stl import BUFFER_SIZE
  File "/media/files/выжигание/bCNC-master/lib/stl/stl.py", line 6, in <module>
    import enum
ImportError: No module named enum
Traceback (most recent call last):
  File "./bCNC.py", line 2598, in <module>
    application = Application(tk)
  File "./bCNC.py", line 189, in __init__
    page = cls(self.ribbon, self)
  File "/media/files/выжигание/bCNC-master/CNCRibbon.py", line 106, in __init__
    self.register()
  File "/media/files/выжигание/bCNC-master/ToolsPage.py", line 1590, in register
    ConfigGroup), (ToolsFrame,))
  File "/media/files/выжигание/bCNC-master/CNCRibbon.py", line 120, in _register
    w = g(self.master._ribbonFrame, self.app)
  File "/media/files/выжигание/bCNC-master/ToolsPage.py", line 1242, in __init__
    image=Utils.icons[tool.icon],
KeyError: 'scale'

**Comment by MARIOBASZ** [comment]
@sfinexer , is GIF file scale 
To use the plugin, unzip, move the scaling file to the plugins folder and the scale file to the icons folder

[scaling.zip](https://github.com/vlachoudis/bCNC/files/2424230/scaling.zip)

**Comment by MARIOBASZ** [comment]
orry, I'm using the version of @harvie: there are problems to run the latest version that contains jpg instead of gif. Look if you replace the gif icon with jpg that I just upload works.
If not, you can modify in line 34 of the plugin and replace with another icon (sorry, I have good intentions, but I am clumsy for many things yet)

[scale.zip](https://github.com/vlachoudis/bCNC/files/2424544/scale.zip)

**Comment by Harvie** [comment]
You have to use .GIF icons everywhere in bCNC. No matter if mine or @vlachoudis !
If you use other formats it might not work on some systems.

**Comment by sfinexer** [comment]
@Harvie but @vlachoudis fixed above correction.
https://github.com/vlachoudis/bCNC/issues/920#issuecomment-405172090
https://github.com/Harvie/bCNC/commit/3f568133003418297c91313fe884cae43dfac19c

**Comment by Harvie** [comment]
@sfinexer i know. i made this commit myself. it's wrong and it's gonna be reverted once my branch is merged. PNG support in tkinter is broken on some platform, so we have to change it back!

**Comment by sfinexer** [comment]
@Harvie Correction is wrong?
Thank you

**Comment by Harvie** [comment]
bCNC will have to go back to GIFs. as done in my branch. Too many people said PNG is broken for them

**Comment by Harvie** [comment]
@vlachoudis is taking too long to merge my fixes, which leads to this confusion.

**Comment by MARIOBASZ** [comment]
I have added fields to avoid parts that are too slow in milling: From Z safe to Z before contact, speed is G0. From there to Z surface Z before contact in% (so that the impact of the tip of the endmill does not suffer). Determine hard crust in which the speed will be hard crust Z feed%

**Comment by MARIOBASZ** [comment]
image

**Comment by Harvie** [comment]
> I have added fields to avoid parts that are too slow in milling: From Z safe to Z before contact, speed is G0. From there to Z surface, speed Z before contact in% (so that the impact of the tip of the endmill does not suffer). Determine hard crust in which the speed will be hard crust Z feed%

In my opinion this is just wrong. If you need this, then you should just lower the "safe Z". it will be even faster!

**Comment by MARIOBASZ** [comment]
You can also use z before contact as relative height from where you are currently (suppose you are flattening with pocket operation in successive passes). By modifying the code, you could upload between passes from current z to z before contact instead of secure z. It may be of no interest to everyone except me.

**Comment by Harvie** [comment]
You can crash te clamps at full speed this way.

**Comment by sfinexer** [comment]
@MARIOBASZ update https://github.com/MARIOBASZ/bCNC
for icons https://github.com/MARIOBASZ/bCNC/tree/master/icons  to gif pleas.

**Comment by sfinexer** [comment]
@MARIOBASZ 
Nothing occurs:
screenshot_20180929_142903

Y0X0Z0
G1F1000
X-20F1000
Z-10F1000
Z0F1000
Y20F1000
Z-2F1000
X-10F1000
Y10F1000
X-20F1000
Y20F1000
Z0F1000
Y0X0F1000

**Comment by MARIOBASZ** [comment]
did not find any error in the uploaded file, I just updated it (I think, I use eurtocarme with the use of Github)
Try on my pc and it works:
image

 upload the compressed plugin
[scaling.zip](https://github.com/vlachoudis/bCNC/files/2430888/scaling.zip)

**Comment by sfinexer** [comment]
@MARIOBASZ 
Well, if one block then everything works.
But in this case the code optimization option does not work.
And if you select several blocks, then the last selected one works.
[test.cnc.zip](https://github.com/vlachoudis/bCNC/files/2430894/test.cnc.zip)

**Comment by MARIOBASZ** [comment]
I had already realized that it works one block at a time, you have to modify it to take several blocks. The original Driller plugin on which I initiate moficaciones, for another utility and on the latter I made the current one, can select several blocks, I have to remember to change it !! I know it's something simple, if you modify it, let me know so I correct it too

**Comment by MARIOBASZ** [comment]
I'm not in my computer, if I remember correctly, if several blocks are selected, driller works but unites them and that's why I made a modification, when I knew how to generate the result on each particular block. I have a lot of ignorance of python and programming in general, on the fly I'm learning, and even then I can contribute!  

Provisional solution until the scaling works on several blocks: use join?

**Comment by Harvie** [comment]
```
		for bid in app.editor.getSelectedBlocks():
			if len(app.gcode.toPath(bid)) < 1: continue

			#nblock = Block("flat "+app.gcode[bid].name())
			#for i in app.gcode[bid]:
			#	nblock.append(re.sub(r"\s?z-?[0-9\.]+","",i))
			#blocks.append(nblock)

			eblock = Block("flat "+app.gcode[bid].name())
			eblock = app.gcode.fromPath(app.gcode.toPath(bid)[0],eblock)
			blocks.append(eblock)
```

**Comment by sfinexer** [comment]
@Harvie There's not so simple as you wrote, there's a few extra G-code commands you need to clean.
There or an exit from "for bid in selectedBlock" early occurs or other.

**Comment by Harvie** [comment]
then just do the cleanup before it calls frompath.

**Comment by MARIOBASZ** [comment]
I suppose for any of you two it's going to be simple. At some point I was going to take the form of selecting which I use @harvie in trochoidal, but since I wanted to work on the 3 domensiones instead of two, and at some point I was stuck, I continued with what I had copied from the driller plugin.
For me the ideal is that it happens as when one does profile operation on different blocks: each block is accompanied by its corresponding modification.

**Comment by sfinexer** [comment]
I completely agree.
I today merged all the branches into one, for simplicity of development. https://github.com/vlachoudis/bCNC/network
I think it would be advisable to take the current branch https://github.com/Sfinexer/bCNC
Not all changes from @Harvie, @MARIOBASZ, @Sfinexer

**Comment by MARIOBASZ** [comment]
I appreciate if you can send in zip as modified, or I can see the changes in your repository

**Comment by MARIOBASZ** [comment]
I have made modifications after Harvie, can you control that my possible clumsiness does not generate conflicts?
I do not know how to update my fork from the one of harvie, I do it from the base of vlachoudis, and even so I do not know if there could be an error because I uploaded changes, then I went up to cancel changes, ... please control that what I have done will not generate conflicts

**Comment by sfinexer** [comment]
@MARIOBASZ I did it, there is the latest version of your change.

**Comment by Harvie** [comment]
> I appreciate if you can send in zip as modified, or I can see the changes in your repository

@MARIOBASZ you really should STOP using files to copy changes between forks. that's a mess. you should always do git pull if you want changes from other repositories!

**Comment by MARIOBASZ** [comment]
I tried to update my repositorio from yours

**Comment by MARIOBASZ** [comment]
I tríes to update my repositorio from yours r

**Comment by MARIOBASZ** [comment]
hey, can you tell me how to work from gcode directly?
Open possibilities!
Among them, I'm interested if you can add some kind of marker in trochoid path for adaptive cutting (something like when you recognize it's an island, or "cut here" to separate joined blocks) instead of recognizing a very small artificial movement of advance and retreat to start decreasing the radio of the trochoids, which is how I could implement it

**Comment by Harvie** [comment]
https://github.com/vlachoudis/bCNC/pull/982#issuecomment-425664577

**Comment by MARIOBASZ** [comment]
I upload my version of trochoidal cut in 3d, to share them.
It is not time yet to ask for the merger, I want to confirm zero errors in specific cases.
You can follow a 3D route, either on a ramp or as a route with steps of different heights in its path. It descends in helix, extends in spiral to the diameter of the trocoid, diameter of adaptive trochoids in corners (working on the resulting path of trochoid path), retracement in a straight line if it is a return along the same path
plugin
[Trochoidal_3D.zip](https://github.com/vlachoudis/bCNC/files/2431628/Trochoidal_3D.zip)
icon
[trochoidal.zip](https://github.com/vlachoudis/bCNC/files/2431630/trochoidal.zip)

**Comment by MARIOBASZ** [comment]
Example: 

[Trochoidal Example.zip](https://github.com/vlachoudis/bCNC/files/2431674/Trochoidal.Example.zip)

Parameters for trochoid path in the AadaptCorner figure:
Cut Diam: 4mm
selected bit: null - bit loaded: 2.5mm (if null, select the loaded bit)
minimal trochoid in% tool: 10%

image

PARAMETERS:
trochoid advance:1.0
Type of slyce: Soft Arc
Clockwise : true
Helical descent:25
splice steps every 360°: 12
Manual setting: false

Material setting:
Z distance before contact: 1
Z feed before contac %t: 80
Hard crust thikness: 1.0
Hard crust Z feed %::70

[Trochoidal Example Resuts.zip](https://github.com/vlachoudis/bCNC/files/2431681/Trochoidal.Example.Resuts.zip)

image

**Comment by MARIOBASZ** [comment]
I do not realize that diameters in helicoidal descent and trochoids coincide.
Here with manual selection trochoi cut diam = 6mm; endmill 0.7mm
Use a spiral to enlarge the hole
[Trochoidal Example Results2.zip](https://github.com/vlachoudis/bCNC/files/2432078/Trochoidal.Example.Results2.zip)

image

**Comment by MARIOBASZ** [comment]
I modified the code in scalar, now I scale all the selected blocks. So as not to generate conflicts in Github, what steps should I follow? I generate a new commit as up to now?

**Comment by MARIOBASZ** [comment]
I will soon upload changes: select several files; Cancel scale for z greater than z surface.

File: icons/scaling.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/python
```
Comment (Harvie) [inline icons/scaling.py]: Why you put .py file in icons directory????

File: icons/bCNC.ini [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+name.0 = default
+material.0 = MDF
+safe.0 = 3.0
+surface.0 = 0.0
+thickness = 5.0
+
+[Cut]
+current = 0
+n = 1
+name.0 = default
```
Comment (Harvie) [inline icons/bCNC.ini]: Why you commit defaults? I don't think this should be here

File: ToolsPage.py [hunk 3]
```diff
@@ -629,7 +629,11 @@ def __init__(self, master):
 			("comment","str",    "", _("Comment")),
 			("feed",    "mm"  , 10., _("Feed")),
 			("feedz",   "mm"  ,  1., _("Plunge Feed")),
-			("stepz",   "mm"  ,  1., _("Depth Increment"))
+			("stepz",   "mm"  ,  1., _("Depth Increment")),
+			("zretract", "mm" ,   1.0, _("Z Distance before contact")),
+			("feedbeforecontact", "int" ,   80, _("Z Feed before contact %")),
+			("hardcrust", "mm" ,   1.0, _("Hard crust thickness")),
+			("hardcrustfeed",  "int", 70, _("Hard crust Z Feed %"))
```
Comment (Harvie) [inline ToolsPage.py]: i really think that this will be only confusing to users without having real benefits

**Comment by Harvie** [comment]
Hello @MARIOBASZ i would like to help you integrate your work to mainline bCNC.
There were lots of changes. Can you please make new `git clone https://github.com/vlachoudis/bCNC.git bCNC-newdir` and try to integrate your changes in there.

In the first step **i only want the new plugins** you have created. Do you think you will be able to make your plugins working in latest @vlachoudis version?

**Comment by Harvie** [comment]
Dear @MARIOBASZ,
i had to rework your commits to reflect latest @vlachoudis master, so i was able to merge that. Here are the resulting commits:

6a8fbd1a34bef0c89c26c0e42a846f4fb3c0c3c5
c477c32e8b922f3926c70c6c8fa51b6304c1a642
77cb1f4f7bb089bb9a7b6107df81a0430fcf9119
ea3bcfc8e2211ff64b55600d56510d116df3c8c2

BUT your adaptive trochoid have bug in it:

```
Traceback (most recent call last):
  File "/home/harvie/Temp/bCNC/bCNC/Utils.py", line 438, in __call__
    return self.func(*args)
  File "/home/harvie/Temp/bCNC/bCNC/ToolsPage.py", line 1698, in execute
    self.tools.getActive().execute(self.app)
  File "/home/harvie/Temp/bCNC/bCNC/plugins/TrochoidPath.py", line 69, in execute
    app.trochprofile_bcnc(self["CutDiam"], direction, self["Offset"], self["Overcut"], self["Adaptative"], cornerradius, name)
  File "/home/harvie/Temp/bCNC/bCNC/__main__.py", line 1848, in trochprofile_bcnc
    msg = self.trochprofile_bcnc(blocks, ofs*sign, overcut, adaptative, adaptedRadius, name)
  File "/home/harvie/Temp/bCNC/bCNC/__main__.py", line 1823, in trochprofile_bcnc
    adaptedRadius = float(adaptedRadius)
TypeError: float() argument must be a string or a number
```

Please do fresh clone of @vlachoudis master branch and continue your work in there, then force push to your github master.

**Pull request closed without merging**